Specific methods for the GCrypt based OKP key generation operation.

javifernandez opened the pull request:
#### bc83c7f25f797f7b4a0cfc8c0bcc0a05e6425361
<pre>
Specific methods for the GCrypt based OKP key generation operation.
<a href="https://bugs.webkit.org/show_bug.cgi?id=261106">https://bugs.webkit.org/show_bug.cgi?id=261106</a>

Reviewed by Žan Doberšek.

This PR defines standalone metods methos for the key generation
operation of the Ed25519 and X25519 algorithms. This approach makes
easier to implement other algorithms that use the OKP key format
as well (eg Ed488 and X488)

It&apos;s just a refactoring with no behavior change, hence we don&apos;t need
to add new tests.

* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::gcryptGenerateEd25519Keys):
(WebCore::gcryptGenerateX25519Keys):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::generateJwkX const):

Canonical link: <a href="https://commits.webkit.org/267621@main">https://commits.webkit.org/267621@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/47a01b97de42a8c67f88276fac92e77b95a247dc

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17216 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17542 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18041 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19005 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16116 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17409 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20816 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17684 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/18307 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17420 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17763 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14955 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19823 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15003 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15646 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/22330 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16002 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15813 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20150 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16405 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13920 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15557 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4105 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19927 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16236 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -63,52 +63,77 @@ static bool supportedAlgorithmIdentifier(CryptoAlgorithmIdentifier keyIdentifier
     return false;
 }
 
+Vector<uint8_t> nine {
```
This should be a static variable.

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -63,52 +63,77 @@ static bool supportedAlgorithmIdentifier(CryptoAlgorithmIdentifier keyIdentifier
 }
 
-std::optional<CryptoKeyPair> CryptoKeyOKP::platformGeneratePair(CryptoAlgorithmIdentifier identifier, NamedCurve namedCurve, bool extractable, CryptoKeyUsageBitmap usages)
+static void gcryptGenerateEd25519Keys(Vector<uint8_t>&& publicKeyData, Vector<uint8_t>&& privateKeyData)
```
This isn't correct usage of rvalue references, but I would recommend a different interface for these two functions:

```
static std::optional<std::pair<Vector<uint8_t>, Vector<uint8_t>>> gcryptGenerateEd25519Keys();
```

This way you return std::nullopt on any error that might occur during generation, and otherwise you return the pair of keys.

javifernandez replied:
Done

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -63,52 +63,77 @@ static bool supportedAlgorithmIdentifier(CryptoAlgorithmIdentifier keyIdentifier
+    // private key is just 32 random bytes
+    PAL::GCrypt::Handle<gcry_mpi_t> mpi(gcry_mpi_new(256));
+    gcry_mpi_randomize(mpi, 256, GCRY_STRONG_RANDOM);
+    auto q = mpiData(mpi);
```
With the new function interface, this would be validated and nullopt returned if for some reason MPI data retrieval failed.

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -63,52 +63,77 @@ static bool supportedAlgorithmIdentifier(CryptoAlgorithmIdentifier keyIdentifier
+    auto q = mpiData(mpi);
+
+    // public key being X25519(a, 9), as defined in [RFC7748], section 6.1.
+    auto d = X25519(*q, WebCore::CryptoKeyOKPImpl::nine);
```
Similarly, this would be tested and nullopt returned if the operation failed, otherwise the newly-constructed std::pair would be returned.

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -63,52 +63,77 @@ static bool supportedAlgorithmIdentifier(CryptoAlgorithmIdentifier keyIdentifier
+
+std::optional<CryptoKeyPair> CryptoKeyOKP::platformGeneratePair(CryptoAlgorithmIdentifier identifier, NamedCurve namedCurve, bool extractable, CryptoKeyUsageBitmap usages)
+{
+    Vector<uint8_t> publicKeyData, privateKeyData;
```
This is then `std::optional<std::pair<Vector<uint8_t>, Vector<uint8_t>>> keyPair;'. By default it's nullopt.

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -63,52 +63,77 @@ static bool supportedAlgorithmIdentifier(CryptoAlgorithmIdentifier keyIdentifier
+        break;
+    case NamedCurve::X25519:
+        gcryptGenerateX25519Keys(WTFMove(publicKeyData), WTFMove(privateKeyData));
+        break;
```
For any of these, `keyPair` is assigned the return value of the corresponding function.

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -63,52 +63,77 @@ static bool supportedAlgorithmIdentifier(CryptoAlgorithmIdentifier keyIdentifier
 
-    auto privateKeyData = mpiData(dMpi);
-    if (!privateKeyData || privateKeyData->size() != 32)
+    if (!(publicKeyData.size() == 32 && privateKeyData.size() == 32))
```
After that switch statement, the `keyPair` variable should be non-null -- it's supported by logic, but you can still assert it. Then you can set up references to the keyPair Vectors:
```
    ASSERT(!!keyPair);
    auto& publicKeyData = keyPair->first;
    auto& privateKeyData = keyPair->second;
```

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -91,24 +95,55 @@ std::optional<CryptoKeyPair> CryptoKeyOKP::platformGeneratePair(CryptoAlgorithmI
 
-    auto privateKeyData = mpiData(dMpi);
-    if (!privateKeyData || privateKeyData->size() != 32)
+    ASSERT(!!keyPair);
```
Sorry, I was completely wrong here. Of course keyPair can still be nullopt, if either function failed and returned nullopt. Please fix before landing.

javifernandez replied:
oh, right ! sorry for not noticing it before. 
Will fix it before landing.

zdobersek commented on Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
```diff
@@ -91,24 +95,55 @@ std::optional<CryptoKeyPair> CryptoKeyOKP::platformGeneratePair(CryptoAlgorithmI
-            && !CryptoConstants::matches(publicKeyData->data(), 1, CryptoConstants::s_x25519UncompressedFormatLeadingByte))) {
+    auto q = mpiData(qMpi);
+    auto d = mpiData(dMpi);
+    if (UNLIKELY((!q || !d)))
```
Nit: I think one pair of parentheses can be removed.

javifernandez replied:
true, thanks

javifernandez committed:
Specific methods for the GCrypt based OKP key generation operation.
https://bugs.webkit.org/show_bug.cgi?id=261106

Reviewed by Žan Doberšek.

This PR defines standalone metods methos for the key generation
operation of the Ed25519 and X25519 algorithms. This approach makes
easier to implement other algorithms that use the OKP key format
as well (eg Ed488 and X488)

It's just a refactoring with no behavior change, hence we don't need
to add new tests.

* Source/WebCore/crypto/gcrypt/CryptoKeyOKPGCrypt.cpp:
(WebCore::gcryptGenerateEd25519Keys):
(WebCore::gcryptGenerateX25519Keys):
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::generateJwkX const):

Canonical link: https://commits.webkit.org/267621@main
